### PR TITLE
Move exp pipe out of numeric

### PR DESCRIPTION
**Pull request by Rhaea**
Finally able to do this with ease. Moved the exponentiation pipe to its own interface in order to remove it from the most basic numeric operations. It never belonged in numeric. I implemented a default version which generates it online if it is not overridden. 
I also would like if we could somehow move random bit and random element from numeric, but this seems not quite as simple and may not be possible. 
A better name for the new interface would not hurt either :) - "PreprocessedValues" seems not quite good.

### CODE REVIEW

File: core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/PreprocessedValues.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package dk.alexandra.fresco.framework.builder.numeric;
+
+import dk.alexandra.fresco.framework.DRes;
+import dk.alexandra.fresco.framework.value.SInt;
+import java.util.List;
+
+/**
+ * Handles values needed internally within FRESCO which preferably should be preprocessed, but a
+ * default implementation could handle it online. In other words: protocol suites are encouraged to
+ * override the default implementation when implementing {@link BuilderFactoryNumeric}.
```
Comment (pffrandsen) [inline core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/PreprocessedValues.java]: "... override the default ... " - which you are neglecting in Spdz - see comment

File: suite/spdz/src/main/java/dk/alexandra/fresco/suite/spdz/SpdzBuilder.java [hunk 2]
```diff
@@ -36,6 +38,18 @@ public BasicNumericContext getBasicNumericContext() {
   }
 
   @Override
+  public PreprocessedValues createPreprocessedValues(ProtocolBuilderNumeric protocolBuilder) {
+    return new PreprocessedValues() {
```
Comment (pffrandsen) [inline suite/spdz/src/main/java/dk/alexandra/fresco/suite/spdz/SpdzBuilder.java]: This should read:
  return new DefaultPreprocessedValues() {

File: core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/DefaultPreprocessedValues.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  
+  public DefaultPreprocessedValues(ProtocolBuilderNumeric builder) {
+    this.builder = builder;
+  }
+  
+  @Override
+  public DRes<List<DRes<SInt>>> getExponentiationPipe(int pipeLength) {
+    return builder.seq(b -> {
+      DRes<SInt> r = b.numeric().randomElement();
+      DRes<SInt> r_inv = b.advancedNumeric().invert(r);
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/DefaultPreprocessedValues.java]: checkstyle violation

File: core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/DefaultPreprocessedValues.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import dk.alexandra.fresco.framework.DRes;
+import dk.alexandra.fresco.framework.value.SInt;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Default implementation which should only be used if for some reason preprocessing is not
+ * possible.
+ * 
+ * @author Kasper Damgaard
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/DefaultPreprocessedValues.java]: Please do not use the author tag. We are not using it consistently, and potentially, many authors will be involved in a file. If you want to see who wrote code there is git for that.

File: core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/DefaultPreprocessedValues.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  public DRes<List<DRes<SInt>>> getExponentiationPipe(int pipeLength) {
+    return builder.seq(b -> {
+      DRes<SInt> r = b.numeric().randomElement();
+      DRes<SInt> r_inv = b.advancedNumeric().invert(r);
+      LinkedList<DRes<SInt>> list = new LinkedList<>();
+      list.add(r_inv);
+      list.add(r);
+      return () -> new IterationState(2, list);
+    }).whileLoop(
+      (state) -> state.round <= pipeLength+1,
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/DefaultPreprocessedValues.java]: checkstyle violation

File: core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/DefaultPreprocessedValues.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      return () -> new IterationState(2, list);
+    }).whileLoop(
+      (state) -> state.round <= pipeLength+1, 
+      (seq, state) -> {      
+        DRes<SInt> r = state.value.get(1);
+        DRes<SInt> last = state.value.getLast();
+        state.value.add(seq.numeric().mult(last, r));
+        state.round++;
+        return state;
+    }).seq((seq, state) -> {
```
Comment (GuutBoy) [inline core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/DefaultPreprocessedValues.java]: Is there a problem with indentation here?

**Pull request merged**